### PR TITLE
Add initial support for controller charms (local only for now)

### PR DESCRIPTION
**Pull request by wallyworld**
Introduce initial support for using a controller charm. The charm will be used to relate to a new dashboard charm.
This initial PR wires up support for specifying a local controller charm at bootstrap to enable development to occur.

Restrictions on the controller charm are:
- it must be called "juju-controller"
- it cannot be deployed manually
- it cannot be removed or have units added/removed

Eventually add/remove unit will replace enable-ha, but there's a lot to do to get there.

## QA steps

Create a local charm called "juju-controller"

```
juju bootstrap lxd --controller-charm /path/to/charm --no-default-model
juju status
```
status shows that there's a controller app with one unit

juju add-unit controller -> error
juju remove-unit controller/0 -> error
juju remove-application controller -> error

### CODE REVIEW

**Comment by wallyworld** [comment]
[deploy local controller charm on bootstrap](https://trello.com/c/e9Jx7GjA/1329-deploy-local-controller-charm-on-bootstrap)

File: go.mod [hunk 1]
```diff
@@ -120,6 +120,7 @@ require (
 	gopkg.in/httprequest.v1 v1.2.1
 	gopkg.in/ini.v1 v1.10.1
 	gopkg.in/juju/blobstore.v2 v2.0.0-20160125023703-51fa6e26128d
+	gopkg.in/juju/charm.v6 v6.0.0-20190729113111-40ffcf7d10e5
```
Comment (SimonRichardson) [inline go.mod:123]: This is the wrong charm. Probably want to do a `go mod tidy`

**Comment by benhoyt** [comment]
Haven't reviewed the code yet, but just had a play with it. I forgot that local charms needed to be zipped up, and so it error'd out, but quite late in the process (after the bootstrap was basically complete). I wonder if we can validate the bootstrap charm beforehand?

**Output pointing `--controller-charm` to a directory:**

```
$ juju bootstrap lxd --controller-charm ~/charms/bentest --no-default-model
Creating Juju controller "localhost-localhost" on lxd/localhost
Looking for packaged Juju agent version 3.0-beta1 for amd64
No packaged binary found, preparing local Juju agent binary
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-c1cd3e-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
No available Juju Dashboard archives found
Waiting for address
Attempting to connect to 10.20.46.51:22
Connected to 10.20.46.51
ERROR failed to bootstrap model: read /home/ben/charms/bentest: is a directory
```

**Output pointing `--controller-charm` to the metadata file:**

```
$ juju bootstrap lxd --controller-charm ~/charms/bentest/metadata.yaml --no-default-model
Creating Juju controller "localhost-localhost" on lxd/localhost
Looking for packaged Juju agent version 3.0-beta1 for amd64
No packaged binary found, preparing local Juju agent binary
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-677d4b-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
No available Juju Dashboard archives found
Waiting for address
Attempting to connect to 10.20.46.247:22
Connected to 10.20.46.247
Running machine configuration script...
Cloud-init v. 20.3-2-g371b392c-0ubuntu1~20.04.1 running 'init-local' at Wed, 06 Jan 2021 23:01:34 +0000. Up 1.44 seconds.

[... hundreds of cloud-init log lines trimmed ...]

2021-01-06 23:05:55 INFO juju.state addmachine.go:498 new machine "0" has preferred addresses: private "local-cloud:10.20.46.247", public "local-cloud:10.20.46.247"
2021-01-06 23:05:59 WARNING juju.cmd.jujud bootstrap.go:375 cannot set up Juju Dashboard: cannot fetch Dashboard info: Dashboard metadata not found
ERROR deploying the controller charm: cannot store controller charm at "/var/lib/juju/charms/controller.charm": invalid charm archive: zip: not a valid zip file
ERROR failed to bootstrap model: subprocess encountered error code 1
```

**As expected, it worked fine with a zip file:**

```
$ juju bootstrap lxd --controller-charm ~/charms/bentest/bentest.zip --no-default-model
Creating Juju controller "localhost-localhost" on lxd/localhost
Looking for packaged Juju agent version 3.0-beta1 for amd64
No packaged binary found, preparing local Juju agent binary
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-804318-0 (arch=amd64)          
Installing Juju agent on bootstrap instance
No available Juju Dashboard archives found
Waiting for address
Attempting to connect to 10.20.46.147:22
Connected to 10.20.46.147
Running machine configuration script...
Bootstrap agent now started
Contacting Juju controller at 10.20.46.147 to verify accessibility...

Bootstrap complete, controller "localhost-localhost" is now available
Controller machines are in the "controller" model

Now you can run
	juju add-model <model-name>
to create a new model to deploy workloads.

$ juju status
Model       Controller           Cloud/Region         Version      SLA          Timestamp
controller  localhost-localhost  localhost/localhost  3.0-beta1.1  unsupported  12:59:01+13:00

App         Version  Status   Scale  Charm            Store  Rev  OS      Message
controller           unknown      1  juju-controller  local    0  ubuntu  

Unit           Workload  Agent  Machine  Public address  Ports  Message
controller/0*  unknown   idle   0        10.20.46.147           

Machine  State    DNS           Inst id        Series  AZ  Message
0        started  10.20.46.147  juju-804318-0  focal       Running
```

File: apiserver/facades/client/application/application.go [hunk 2]
```diff
@@ -1633,6 +1634,21 @@ func (api *APIBase) AddUnits(args params.AddApplicationUnits) (params.AddApplica
...
+	app, err := api.backend.Application(args.ApplicationName)
+	if err != nil {
+		return params.AddApplicationUnitsResults{}, errors.Trace(err)
+	}
+	ch, _, err := app.Charm()
+	if err != nil {
+		return params.AddApplicationUnitsResults{}, errors.Trace(err)
+	}
+	if ch.Meta().Name == bootstrap.ControllerCharmName {
+		return params.AddApplicationUnitsResults{}, errors.NotSupportedf("add units to the controller application")
```
Comment (benhoyt) [inline apiserver/facades/client/application/application.go:1649]: Should this read "add*ing* units ... not supported"?

File: apiserver/facades/client/application/application.go [hunk 3]
```diff
@@ -1777,6 +1796,26 @@ func (api *APIBase) DestroyUnit(args params.DestroyUnitsParams) (params.DestroyU
...
+				return nil, errors.Trace(err)
+			}
+			ch, _, err = app.Charm()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			appCharms[appName] = ch
+		}
+		if ch.Meta().Name == bootstrap.ControllerCharmName {
+			return nil, errors.NotSupportedf("remove units from the controller application")
```
Comment (benhoyt) [inline apiserver/facades/client/application/application.go:1816]: Similar: should it read "remov*ing* units ... not supported"?

File: go.mod (hunk 1 above)
Comment (wallyworld) [reply go.mod]: doh, thanks

File: cmd/juju/commands/bootstrap.go [hunk 4]
```diff
@@ -326,6 +328,7 @@ func (c *bootstrapCommand) SetFlags(f *gnuflag.FlagSet) {
 	f.BoolVar(&c.noSwitch, "no-switch", false, "Do not switch to the newly created controller")
 	f.BoolVar(&c.Force, "force", false, "Allow the bypassing of checks such as supported series")
 	f.BoolVar(&c.noHostedModel, "no-default-model", false, "Do not create a default model")
+	f.StringVar(&c.ControllerCharmPath, "controller-charm", "", "Path to a locally built controller.charm")
```
Comment (benhoyt) [inline cmd/juju/commands/bootstrap.go:331]: "controller.charm" -> "controller charm"?

File: cmd/juju/commands/bootstrap.go [hunk 5]
```diff
@@ -354,6 +357,13 @@ func (c *bootstrapCommand) Init(args []string) (err error) {
 		return errors.New("--db-snap-asserts requires --db-snap")
 	}
 
+	if c.ControllerCharmPath != "" {
+		_, err := c.Filesystem().Stat(c.ControllerCharmPath)
```
Comment (benhoyt) [inline cmd/juju/commands/bootstrap.go:361]: Is this where we could do a bit of additional validation on the charm file, e.g., ensure `ReadCharmArchive()` returns success?

File: cmd/jujud/agent/bootstrap.go [hunk 6]
```diff
@@ -382,6 +390,102 @@ func (c *BootstrapCommand) Run(_ *cmd.Context) error {
...
+		},
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	u, err := app.AddUnit(state.AddUnitParams{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return u.AssignToMachine(m)
```
Comment (benhoyt) [inline cmd/jujud/agent/bootstrap.go:486]: I dunno, this seems a touch heavy on the single-letter pedal to me. Maybe `unit` and `machine`?

**Review by benhoyt — APPROVED** [review]

File: cmd/jujud/agent/bootstrap.go (hunk 6 above)
Comment (wallyworld) [reply cmd/jujud/agent/bootstrap.go:486]: We use this in a few places so was going for consistency.

File: cmd/juju/commands/bootstrap.go (hunk 5 above)
Comment (wallyworld) [reply cmd/juju/commands/bootstrap.go:361]: Yup, added extra checks

File: cmd/juju/commands/bootstrap.go (hunk 4 above)
Comment (wallyworld) [reply cmd/juju/commands/bootstrap.go:331]: fixed

File: apiserver/facades/client/application/application.go (hunk 2 above)
Comment (wallyworld) [reply apiserver/facades/client/application/application.go:1649]: fixed

File: apiserver/facades/client/application/application.go (hunk 3 above)
Comment (wallyworld) [reply apiserver/facades/client/application/application.go:1816]: fixed

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**